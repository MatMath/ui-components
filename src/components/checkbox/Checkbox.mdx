---
name: Checkbox
menu: Components
---

import { Playground, Props } from 'docz';
import { Checkbox } from './Checkbox';
import { CheckboxDemo } from './CheckboxDemo';

# Checkbox

A checkbox component

## Properties

<Props of={Checkbox} />

## Usage

### Basic usage

Examples below show the `Checkbox` in its checked and unchecked states, both in enabled and disabled modes:

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around'
  }}
>
  <Checkbox readOnly />
  <Checkbox defaultChecked checked readOnly />
  <Checkbox disabled readOnly />
  <Checkbox disabled defaultChecked checked readOnly />
</Playground>

The example below shows an interactive `Checkbox`.

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around'
  }}
>
  <CheckboxDemo />
</Playground>
