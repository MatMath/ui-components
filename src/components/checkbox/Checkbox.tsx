import * as React from 'react';
import { getClassNames } from '@utility/cssUtils';
import styles from './Checkbox.module.scss';

const CheckboxContainer = ({
  className,
  children
}: React.InputHTMLAttributes<HTMLInputElement>) => {
  return (
    <label className={getClassNames(styles.checkboxContainer, className || '')}>
      {children}
    </label>
  );
};

const HiddenCheckbox = ({
  checked = false,
  ...props
}: React.InputHTMLAttributes<HTMLInputElement>) => {
  return (
    <input
      className={styles.hiddenCheckBox}
      type='checkbox'
      checked={checked}
      {...props}
    />
  );
};

const StyledCheckbox = ({
  checked = false,
  disabled = false,
  children
}: React.InputHTMLAttributes<HTMLInputElement>) => {
  const className = getClassNames(styles.styledCheckbox, {
    [styles.checked]: checked,
    [styles.disabled]: disabled
  });
  return <div className={className}>{children}</div>;
};

const Checkmark = () => {
  return (
    <svg
      className={styles.checkmark}
      width='20'
      height='20'
      viewBox='0 0 20 20'
      fill='none'
      xmlns='http://www.w3.org/2000/svg'
    >
      <path
        className='checkmark__check'
        d='M5.5 10L8.45928 13L14.4796 7'
        strokeLinecap='round'
        strokeLinejoin='round'
      />
    </svg>
  );
};

export const Checkbox = ({
  className,
  checked,
  disabled,
  ...props
}: React.InputHTMLAttributes<HTMLInputElement>) => (
  <CheckboxContainer className={className}>
    <HiddenCheckbox checked={checked} {...props} />
    <StyledCheckbox checked={checked} disabled={disabled}>
      <Checkmark />
    </StyledCheckbox>
  </CheckboxContainer>
);
