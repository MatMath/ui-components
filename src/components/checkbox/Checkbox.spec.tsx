import * as React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Checkbox } from './Checkbox';

describe('<Checkbox>', () => {
  describe('Global render', () => {
    it('should render default', () => {
      const { container } = render(<Checkbox />);
      expect(container).toMatchSnapshot();
      expect(screen.getByRole('checkbox')).not.toHaveAttribute('checked');
    });

    it('should render disabled', () => {
      const { container } = render(<Checkbox disabled />);
      expect(container).toMatchSnapshot();
    });

    it('should render checked', () => {
      const { container } = render(<Checkbox checked />);
      expect(container).toMatchSnapshot();
      expect(screen.getByRole('checkbox')).toHaveAttribute('checked');
    });

    it('should be checked on click', () => {
      render(<Checkbox />);
      expect(screen.getByRole('checkbox')).not.toBeChecked();
      userEvent.click(screen.getByRole('checkbox'));
      expect(screen.getByRole('checkbox')).toBeChecked();
    });

    it('should render unchecked and with a label', () => {
      const label = 'Label';
      render(<Checkbox label={label} />);
      expect(screen.getByLabelText(label)).toBeInTheDocument();
    });
  });
});
