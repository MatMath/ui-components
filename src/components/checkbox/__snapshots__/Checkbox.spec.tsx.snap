// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> Global render should render checked 1`] = `
<div>
  <label
    class="checkboxContainer"
  >
    <input
      checked=""
      class="hiddenCheckBox"
      type="checkbox"
    />
    <div
      class="styledCheckbox checked"
    >
      <svg
        class="checkmark"
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 10L9 13L15.0407 7"
          stroke-linecap="round"
        />
      </svg>
    </div>
  </label>
</div>
`;

exports[`<Checkbox> Global render should render default 1`] = `
<div>
  <label
    class="checkboxContainer"
  >
    <input
      class="hiddenCheckBox"
      type="checkbox"
    />
    <div
      class="styledCheckbox"
    >
      <svg
        class="checkmark"
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 10L9 13L15.0407 7"
          stroke-linecap="round"
        />
      </svg>
    </div>
  </label>
</div>
`;

exports[`<Checkbox> Global render should render disabled 1`] = `
<div>
  <label
    class="checkboxContainer disabled"
  >
    <input
      class="hiddenCheckBox"
      disabled=""
      type="checkbox"
    />
    <div
      class="styledCheckbox disabled"
    >
      <svg
        class="checkmark"
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 10L9 13L15.0407 7"
          stroke-linecap="round"
        />
      </svg>
    </div>
  </label>
</div>
`;
